armi/reactor/blocks.py

```diff
@@ -188,7 +188,7 @@ def r(self):
 
         core = self.core
         if core is None:
-            return None
+            return self.getAncestor(lambda o: isinstance(o, Reactor))
```

Comment:
This change only appeared for `Composites` that were not directly attached to the core, but
more acted as siblings to the `Core`. For example, if the spent fuel pool composite was
a child of `Reactor`, and the spent fuel pool contained assemblies which in turn contained blocks,
various HDF serialization methods would error out because Block.core -> None.

One example:
_addHomogenizedNumberDensityParams -> collectBlockNumberDensities ->  getSymmetryFactor
calls self.r.core.spatialGrid.overlapsWhichSymmetryLine

The docstring for Block.r indicates that most calls to self.r are to obtain the core
and may be better off just calling Block.core. But, here exists a case where a Block
may not have a Core as the parent (e.g., ex-core structures that physically reside
in the reactor) but still needs to access, for some reason, the spatial grid.

A deeper issue is that this call stack is trying to find the spatial grid for a block
assuming the block is on the Core.spatialGrid. Not sure how to square that. Maybe
the call for symmetry line tries to find a parent with a spatialGrid?

Reply:
Perhaps this could use a one-line comment explaining why?

Reply:
@ntouran  I can't tell, will this be confusing to users?  They could ask for `b.r` and get something that is not the `Reactor`.

Reply:
> They could ask for b.r and get something that is not the Reactor.

That actually would have been the previous case if you had spun up a Reactor that contained a `Core` and some sibling `Composite` that also contained `Block` s and `Assembly` s. Because this sibling isn't the `Core`, blocks assigned have a non-None `self.core` attribute from
https://github.com/terrapower/armi/blob/a3f23ed01538aaaeb1dc58b5bf158def918901ef/armi/reactor/blocks.py#L165-L170

So here, for the blocks in the `Core`-adjacent `Composite`, their parent tree looks like `Block` -> `Assembly` -> `Sibling` -> `Reactor`. Then, if we did `b.core` we get `None,` which makes `b.r` return `None` without this change.